src/NHibernate/Context/MapBasedSessionContext.cs

```diff
@@ -20,30 +23,35 @@ protected override ISession Session
 		{
 			get
 			{
-				IDictionary map = GetMap();
-				if (map == null)
-				{
-					return null;
-				}
-				else
-				{
-					return map[_factory] as ISession;
-				}
+				return GetConcreteMap()?.GetOrAdd(_factory, default(ISession));
```

Comment:
A `GetOrDefault` would have been enough here, but as it does not exists, `GetOrAdd` is the way to go. This potential add will not harm: having no entry or having `null` is equivalent. And its semantics ensures it will not override a concurrent item set from the setter.

`default(ISession)` required instead of `null` due to overload ambiguity.